### PR TITLE
Try: Replace classnames with clsx (outdated)

### PR DESCRIPTION
**Pull request by DaniGuardiola**
**Important:** superseded by #61138

## What?

This PR attempts to replace classnames with clsx across Gutenberg, which is a faster and lighter alternative.

@tyxla already did this in calypso: https://github.com/Automattic/wp-calypso/pull/89843

## How?

It's a drop-in replacement. All `classnames/dedupe` instances were also safe to replace, except for one case where [an obscure feature](https://www.npmjs.com/package/classnames#alternate-dedupe-version) was being used - that was trivially replaced with equivalent code (see `removeAspectRatioClasses` commit for details).

## Size gains

With `classnames` and `classnames/dedupe` we added 1.14kb (minified + gzipped) to the bundle. See:https://bundlejs.com/?q=classnames%402.5.1%2Cclassnames%402.5.1%2Cclassnames%402.5.1%2Fdedupe%2Cclassnames%402.5.1%2Fdedupe&treeshake=%5B*%5D%2C%5B%7B+default+%7D%5D%2C%5B*%5D%2C%5B%7B+default+as+d2+%7D%5D

With clsx, we only add 241B (minified and gzipped). See: https://bundlejs.com/?q=clsx%402.1.1&treeshake=%5B*%5D

That's about a kylobyte of savings.

### CODE REVIEW

**Comment by DaniGuardiola** [comment]
Lots of CI failures that seems completely unrelated to the changes (basically a grep replace of classnames -> clsx.

Anyone know why this is happening? Is this just a side effect of touching a lot of repo-wide code?

**Comment by tyxla** [comment]
> Lots of CI failures that seems completely unrelated to the changes (basically a grep replace of classnames -> clsx.
> 
> Anyone know why this is happening? Is this just a side effect of touching a lot of repo-wide code?

I'd suggest double-checking if it's really unrelated. There seems to be some relation at first glance:

Screenshot 2024-04-25 at 12 40 39

**Comment by DaniGuardiola** [comment]
@tyxla most do seem unrelated but I also stumbled upon that one. I'm puzzled about that instance because I can't find what's being referenced there. Literally grepping for "classnames" case insensitive doesn't find any more relevant references in code. I feel like I'm missing something 🤔

**Comment by tyxla** [comment]
> @tyxla most do seem unrelated but I also stumbled upon that one. I'm puzzled about that instance because I can't find what's being referenced there. Literally grepping for "classnames" case insensitive doesn't find any more relevant references in code. I feel like I'm missing something 🤔

Have you tried rebasing to the latest `trunk`? Seems like a `classnames` import was introduced there literally 40 mins ago in https://github.com/WordPress/gutenberg/pull/60894.

**Comment by DaniGuardiola** [comment]
@tyxla that was it. Pushing fix. Thanks!

**Comment by DaniGuardiola** [comment]
Ok got some tests related to `removeAspectRatioClasses` failing, probably because of extra spaces in my naive replacement solution. Should be easy to remedy, fix incoming.

File: docs/contributors/code/coding-guidelines.md [hunk 1]
```diff
@@ -39,13 +39,13 @@ Components may be assigned with class names that indicate states (for example, a
 
 **Example:**
 
-Consider again the Notices example. We may want to apply specific styling for dismissible notices. The [`classnames` package](https://www.npmjs.com/package/classnames) can be a helpful utility for conditionally applying modifier class names.
+Consider again the Notices example. We may want to apply specific styling for dismissible notices. The [`clsx` package](https://www.npmjs.com/package/clsx) can be a helpful utility for conditionally applying modifier class names.
 
 ```jsx
-import classnames from 'classnames';
+import clsx from 'clsx';
```
Comment (ellatrix) [inline docs/contributors/code/coding-guidelines.md:45]: Could we create our own class names function and use clx under the hood? That we can recommend using our own function and if we ever need to swap to a different package we don't need to update all these instances and ask plugin devs to change it.

Comment (tyxla) [reply docs/contributors/code/coding-guidelines.md:45]: I like that idea. Where would you expect to see that function live @ellatrix?

Comment (ellatrix) [reply docs/contributors/code/coding-guidelines.md:45]: Depends, do the affected packages have a dependency in common? Maybe `@wordpress/element`?

Comment (ellatrix) [reply docs/contributors/code/coding-guidelines.md:45]: Since it's meant to be used with React components. Not sure if there's other use cases.

Comment (DaniGuardiola) [reply docs/contributors/code/coding-guidelines.md:45]: I don't see much of an advantage to wrapping this package. It's a very simple and standard 200 bytes utility. Easy to swap too if it's ever necessary, especially after this PR where I'm taking care of the only edge case from the previous classnames usage.

Comment (mirka) [reply docs/contributors/code/coding-guidelines.md:45]: I don't think this concerns plugin devs in any way, it's only a change internal to Gutenberg. The only benefit I see would be deduplication of the library code, which is tiny.

Given that we're also now encouraging direct imports of React (#54074), and we have no reason to be prescriptive about what classname utility to use in plugins, I think we can just keep using `clsx` directly?

Comment (DaniGuardiola) [reply docs/contributors/code/coding-guidelines.md:45]: This seems out of the scope of this PR, which is just meant to replace an outdated utility with a modern drop-in replacement. This suggestion is better suited for a separate issue IMO. Appreciate you chiming in though :)

**Comment by DaniGuardiola** [comment]
This is ready for review now. ~Not sure why some PHP lints are failing, since no PHP has been touched 🤷~ Found the issue, some automatic formatting of a PHP file slipped by. Not sure exactly how, but that's been reverted now.

**Review by mirka — COMMENTED** [review]
I think we should include the restricted import [eslint rule](https://github.com/WordPress/gutenberg/blob/41c80e8551294914681b060caded03cfe563c6b6/.eslintrc.js#L39) in this PR so nobody wastes their time figuring out what happened.

File: .eslintrc.js [hunk 2]
```diff
@@ -76,6 +76,11 @@ const restrictedImports = [
 		message:
 			"edit-widgets is a WordPress top level package that shouldn't be imported into other packages",
 	},
+	{
+		name: 'classnames',
```
Comment (tyxla) [inline .eslintrc.js:80]: Thanks for adding an eslint restricted import rule 👍 Should greatly help with the confusion.

**Review by tyxla — APPROVED** [review]
This looks good and works well. 👍 

I'd add some entries to all existing CHANGELOGs of affected packages since it's a change worth mentioning.

Since it's a change that affects the entire repo, perhaps we could wait for a couple of days before merging, in case someone has blocking feedback.

Thanks @DaniGuardiola 🙌

**Comment by youknowriad** [comment]
Is there any reason why the bundle size comment is not visible on this PR, I wanted to check the impact.

I believe this needs to be announced in #core-editor at least to bring some visibility among contributors but other than that, it seems fine to land to me.

File: docs/contributors/code/coding-guidelines.md (hunk 1 above)
Comment (ellatrix) [reply docs/contributors/code/coding-guidelines.md:45]: > Easy to swap too if it's ever necessary

You still need to change all the imports 🙂

Comment (youknowriad) [reply docs/contributors/code/coding-guidelines.md:45]: I also feel like we shouldn't try to abstract this. Say tomorrow React handles this automatically, it won't be a simple library change, we'll have to iteratively remove the usage instead.

**Comment by DaniGuardiola** [comment]
@youknowriad

> Is there any reason why the bundle size comment is not visible on this PR, I wanted to check the impact.

I'm curious too. Could someone more knowledgeable look into it? :)

> I believe this needs to be announced in #core-editor at least to bring some visibility among contributors

Good point 👍 will make sure to do this after merge.

**Comment by youknowriad** [comment]
> I'm curious too. Could someone more knowledgeable look into it? :)

It seems it's because you're using a fork. I saw that you already have 3 PRs merged into the repo so I added you to the Gutenberg team to be able to create branches and merge approved PRs.

If you want to check the bundle size impact before merging here, you'll have to recreate another PR using a branch in the main repo I guess.

**Comment by DaniGuardiola** [comment]
Closing this in favor of #61138 since I've continued work there.

**Pull request closed without merging**